### DISCUSSION TITLE
Node_modules updates and caching

### DISCUSSION THREAD

**Discussion by nikwotton** [vercel/turborepo#8896] [Help]
### Summary

I'm not sure if this belongs as a bug report or is intended behavior so opening a discussion for now. Turborepo says in multiple places that it does not manage dependencies which is fair, I can leave that up to npm. I am attempting to enable caching through turborepo, which on its face seems trivial. I think I've found an issue though with how turbo interacts with dependencies and am looking for the best way forward. 

Let's say I have a basic turbo.json:
```json
{
  "$schema": "https://turbo.build/schema.json",
  "tasks": {
    "tsc": {
      "output": "dist/",
      "cache": true
    }
  }
}
```

Now for the sake of simplicity I have two branches, `main` and `feature/foobar`. Running `turbo tsc` on `main` works fine. The only change on `feature/foobar` vs `main` is deleting an essential dependency from `package.json` so of course running `turbo tsc` on `feature/foobar` is expected to fail. Now I follow these steps:

1. `git checkout main`
2. `rm -rf .turbo/ node_modules/`
3. `npm i`
4. `git checkout feature/foobar`
5. `turbo tsc` - it passes because `node_modules/` has not been updated to reflect the dependencies change
6. `npm i`
7. `turbo tsc`

Step 7 passes because the unintentional passing in step 5 is now cached and as far as turbo is concerned no inputs or outputs have changed, therefore the cache should be used. As a developer I would expect step 7 to fail and not pull from cache. 

Is this a documented use case I've missed?

**Comment by nikwotton** [comment]
Note the best/only solution I've found so far is to add an `installDependencies` script with `package-lock.json` as an input and `node_modules/` as an output and have this be a dependency of all other scripts. It massively increases the size of the `.turbo` directory (for obvious reasons) and seems inefficient but I don't have a better approach yet - really hoping this is a solved problem and someone tells me I'm just missing some flag

**Comment by anthonyshew** [comment]
Cache will be missed for this case according to the default behavior. Both the change to `package.json` and the change to the lockfile should result in a hash change. Can you provide a fuller reproduction?

**Reply by nikwotton** [reply]
@anthonyshew You are correct that when the `package.json` and `package-lock.json` change at step 4, the change causes the cache to miss due to a hash change in step 5. As described above though, step 5 passes because `node_modules` was not updated, then running `npm i` in step 6 does NOT cause a hash change and therefore does not lead to a cache miss at step 7

**Reply by anthonyshew** [reply]
Ah, I'm realize what you're saying now. node_modules is out of sync with your lockfile when you run the build.

Turborepo doesn't hash on the contents of node_modules and doesn't have a mechanism for checking that node_modules is or isn't out of date. Your best bet is likely to do like you're describing, levaraging your package manager to make sure things are synced up. For what it's worth. I _think_ if I were to do this, I probably wouldn't cache it.

**Reply by nikwotton** [reply]
Correct. FWIW What I'm doing now does not involve caching, but I've got `npm i` running as fast I can and am just running it as a dependency of all other tasks - not ideal but it's the best I've got so far unfortunately
